### PR TITLE
Makes GenericRecord to string more readable

### PR DESCRIPTION
**Pull request by sancar**
GenericRecord is printed in json format.

### CODE REVIEW

File: hazelcast/src/test/java/com/hazelcast/internal/serialization/impl/compact/GenericRecordTest.java [hunk 1]
```diff
@@ -41,6 +58,108 @@
...
+        assertTrue(data.isCompact());
+
+        //internal generic record created on the servers on query
+        InternalGenericRecord internalGenericRecord = serializationService.readAsInternalGenericRecord(data);
+        Json.parse(internalGenericRecord.toString());
+
+        //generic record read from a remote instance without classes on the classpath
+        List<String> excludes = Collections.singletonList("example.serialization");
+        FilteringClassLoader classLoader = new FilteringClassLoader(excludes, null);
+        Thread.currentThread().setContextClassLoader(classLoader);
```
Comment (ihsandemir) [inline hazelcast/src/test/java/com/hazelcast/internal/serialization/impl/compact/GenericRecordTest.java:148]: Do you need to revert back the class loader for the thread after this test case finishes?

**Review by ihsandemir — APPROVED** [review]

File: hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/FieldOperations.java [hunk 2]
```diff
@@ -182,6 +228,11 @@ public Object readObject(GenericRecord genericRecord, String fieldName) {
         public void writeFieldFromRecordToWriter(DefaultCompactWriter writer, GenericRecord genericRecord, String fieldName) {
             writer.writeString(fieldName, genericRecord.getString(fieldName));
         }
+
+        @Override
+        public String readAsJsonFormattedField(AbstractGenericRecord record, String fieldName) {
+            return "\"" + record.getString(fieldName) + "\"";
```
Comment (viliam-durina) [inline hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/FieldOperations.java:234]: When it looks like JSON, it probably should be JSON and we should escape the strings. Also field names should be escaped then.

Comment (sancar) [reply hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/FieldOperations.java:234]: Make sense. I wonder if we have anything reusable in our codebase. `JsonWriter` looks like have somethinng but not reusable at the moment.

File: hazelcast/src/test/java/com/hazelcast/internal/serialization/impl/compact/GenericRecordTest.java (hunk 1 above)
Comment (sancar) [reply hazelcast/src/test/java/com/hazelcast/internal/serialization/impl/compact/GenericRecordTest.java:148]: Makes sense to be cautious. I will address this.

File: hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/FieldOperations.java (hunk 2 above)
Comment (sancar) [reply hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/FieldOperations.java]: @viliam-durina Updated the pr accordingly. Can you check again ?

Comment (viliam-durina) [reply hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/FieldOperations.java]: I only glanced, looks ok. But it's quite inefficient. Much better way would be if the escape method directly copied to the StringBuilder, e.g. `public static void escape(String source, StringBuilder target)`

File: hazelcast/src/main/java/com/hazelcast/internal/json/JsonEscape.java [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+package com.hazelcast.internal.json;
+
+import static com.hazelcast.internal.json.JsonWriter.getReplacementChars;
+
+/**
+ * Utility method to escape characters for fields used in a Json.
+ */
+public class JsonEscape {
+
+    public static void writeEscaped(StringBuilder stringBuilder, String string) {
```
Comment (viliam-durina) [inline hazelcast/src/main/java/com/hazelcast/internal/json/JsonEscape.java:26]:
```suggestion
    public static void writeEscaped(StringBuilder target, String source) {
```

File: hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/FieldOperations.java [hunk 4]
```diff
@@ -91,6 +119,13 @@ public void writeFieldFromRecordToWriter(DefaultCompactWriter writer, GenericRec
         public int typeSizeInBytes() {
             return Character.BYTES;
         }
+
+        @Override
+        public void writeJsonFormattedField(StringBuilder stringBuilder, AbstractGenericRecord record, String fieldName) {
+            stringBuilder.append('"');
+            stringBuilder.append(record.getChar(fieldName));
+            stringBuilder.append('"');
```
[the remark this answers is not in the document]
Comment (sancar) [reply hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/FieldOperations.java:127]: Good catch. Correcting it.

File: hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/FieldOperations.java [hunk 5]
```diff
@@ -209,6 +279,26 @@ public int hashCode(GenericRecord record, String fieldName) {
...
+            if (objects == null) {
+                stringBuilder.append("null");
+                return;
+            }
+            stringBuilder.append("[");
+            int objectsSize = objects.length;
+            int j = 0;
+            for (GenericRecord object : objects) {
+                j++;
+                stringBuilder.append(object);
```
[the remark this answers is not in the document]
Comment (sancar) [reply hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/FieldOperations.java:295]: Portable does not allow but Compact allows null. 
I will change this to Arrays.toString. 
And in compactArray , I will check for nulls on the items.

File: hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/FieldOperations.java [hunk 6]
```diff
@@ -398,6 +544,26 @@ public int hashCode(GenericRecord record, String fieldName) {
...
+            if (objects == null) {
+                stringBuilder.append("null");
+                return;
+            }
+            stringBuilder.append("[");
+            int size = objects.length;
+            int j = 0;
+            for (String object : objects) {
+                j++;
+                JsonEscape.writeEscaped(stringBuilder, object);
```
[the remark this answers is not in the document]
Comment (sancar) [reply hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/FieldOperations.java:560]: I am adding a generic method where it handles null items in arrays. I will use single method for all(whether a null is allowed or not)

**Comment by sancar** [comment]
@mdumandag From GenericRecordTest 
```
{"example.serialization.MainDTO": {"b": 113,
"bigDecimal": 12312313,
"bool": true,
"c": "x",
"d": -897543.3678909,
"f": 900.5678,
"i": 56789,
"l": -50992225,
"localDate": "2021-08-31",
"localDateTime": "2021-08-31T09:58:00.753",
"localTime": "09:58:00.753",
"offsetDateTime": "2021-08-31T09:58:00.753+03:00",
"p": {"example.serialization.InnerDTO": {"bb": [0, 1, 2],
"bigDecimals": [12345,123456],
"cc": ["c","h","a","r"],
"dd": [456.456, 789.789, 321.321],
"ff": [0.6543, -3.56, 45.67],
"ii": [9, 8, 7, 6],
"ll": [0, 1, 5, 7, 9, 11],
"localDateTimes": ["2021-08-31T09:58:00.753"],
"localDates": ["2021-08-31","2021-08-31"],
"localTimes": ["09:58:00.752","09:58:00.752"],
"nn": [{"example.serialization.NamedDTO": {"myint": 123,
"name": "name"}},{"example.serialization.NamedDTO": {"myint": 123,
"name": "name"}}],
"offsetDateTimes": ["2021-08-31T09:58:00.753+03:00"],
"ss": [3, 4, 5]}},
"s": -500,
"str": "this is main object created for testing!"}}

```

**Comment by mdumandag** [comment]
Honestly, the output looks pretty unreadable to me. Especially, for big objects like you shared, the lack of the correct level of indentation makes it hard to understand which key belongs to which object. I would vote for making use of JsonObject even if it might cause more litter. @sancar @ihsandemir WDYT?

File: hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/FieldOperations.java [hunk 7]
```diff
@@ -272,6 +357,27 @@ public int hashCode(GenericRecord record, String fieldName) {
...
+                return;
+            }
+            stringBuilder.append("[");
+            int objectsSize = objects.length;
+            int j = 0;
+            for (char object : objects) {
+                j++;
+                JsonEscape.writeEscaped(stringBuilder, object);
+                if (j != objectsSize) {
+                    stringBuilder.append(",");
```
Comment (mdumandag) [inline hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/FieldOperations.java:375]:
```suggestion
                    stringBuilder.append(", ");
```
Minor: To make it consistent with the behavior of the Arrays.toString

File: hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/FieldOperations.java [hunk 8]
```diff
@@ -409,6 +546,12 @@ public Object readObject(GenericRecord genericRecord, String fieldName) {
         public void writeFieldFromRecordToWriter(DefaultCompactWriter writer, GenericRecord genericRecord, String fieldName) {
             writer.writeDecimal(fieldName, genericRecord.getDecimal(fieldName));
         }
+
+        @Override
+        public void writeJsonFormattedField(StringBuilder stringBuilder, AbstractGenericRecord record, String fieldName) {
+            BigDecimal value = record.getDecimal(fieldName);
+            stringBuilder.append(value == null ? "null" : value.toString());
```
Comment (mdumandag) [inline hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/FieldOperations.java:553]:
```suggestion
            stringBuilder.append(record.getDecimal(fieldName));
```
Minor: This has the same effect

File: hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/FieldOperations.java [hunk 9]
```diff
@@ -594,6 +812,16 @@ public GenericRecord readGenericRecordOrPrimitive(GenericRecord genericRecord, S
...
         }
+
+        @Override
+        public void writeJsonFormattedField(StringBuilder stringBuilder, AbstractGenericRecord record, String fieldName) {
+            GenericRecord genericRecord = record.getGenericRecord(fieldName);
+            if (genericRecord == null) {
+                stringBuilder.append("null");
+                return;
+            }
+            stringBuilder.append(genericRecord);
```
Comment (mdumandag) [inline hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/FieldOperations.java:823]:
```suggestion
            stringBuilder.append(record.getGenericRecord(fieldName));
```
Minor: This has the same effect

File: hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/FieldOperations.java [hunk 10]
```diff
@@ -634,4 +868,31 @@ public static FieldTypeBasedOperations fieldOperations(FieldType fieldType) {
...
+        int j = 0;
+        for (T object : objects) {
+            j++;
+            if (object == null) {
+                stringBuilder.append("null");
+            } else {
+                itemWriter.accept(stringBuilder, object);
+            }
+            if (j != objectSize) {
+                stringBuilder.append(",");
```
Comment (mdumandag) [inline hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/FieldOperations.java:893]:
```suggestion
                stringBuilder.append(", ");
```
Minor: To make it consistent with others

File: hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/portable/PortableGenericRecord.java [hunk 11]
```diff
@@ -257,7 +257,11 @@ protected Object getClassIdentifier() {
 
     @Override
     public String toString() {
-        return "PortableGenericRecord:" + super.toString();
+        StringBuilder stringBuilder = new StringBuilder();
+        stringBuilder.append("{ \"Portable\": ");
```
Comment (mdumandag) [inline hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/portable/PortableGenericRecord.java:261]:
```suggestion
        stringBuilder.append("{\"Portable\": ");
```

File: hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/portable/PortableInternalGenericRecord.java [hunk 12]
```diff
@@ -872,4 +872,13 @@ public Object getObject(@Nonnull String fieldName) {
     protected Object getClassIdentifier() {
         return cd;
     }
+
+    @Override
+    public String toString() {
+        StringBuilder stringBuilder = new StringBuilder();
+        stringBuilder.append("{ \"Portable\": ");
```
Comment (mdumandag) [inline hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/portable/PortableInternalGenericRecord.java:879]:
```suggestion
        stringBuilder.append("{\"Portable\": ");
```

**Review by mdumandag — APPROVED** [review]
Approving with minor comments

**Pull request merged**